### ISSUE TITLE
Using join function returns no results when running phpunit with queries over multiple databases

### ISSUE THREAD

**Issue by coxy121** [laravel/framework#24945]
- Laravel Version: 5.6.28
- PHP Version: 7.2.2
- Database Driver & Version: MySQL 5.5.50

### Description:
I am using multiple MySQL databases in my application, and have to run phpunit using MySQL databases. The issues I am experiencing is when using multiple mysql databases with PHPUnit join queries across 2 databases return no data. This only happens when running the tests via phpunit, and works as expected otherwise.

When running the following test in phpunit several queries are run to retrieve the data using the join function. The table linked to each model is stored in a seperate database and in the model class the connection is defined, as either mysql or mysql2.
 
```
public function testBasicTest()
    {
         $user = factory(\App\User::class)->create();

        factory(\App\UserType::class)->create(['user_id' => $user->id]);

        $response = $this->get('/');

        $response->assertStatus(200);
    }
```

This would return the collection as expected with the relationship data loaded
```
\App\User::with('type')->get()
```

However this returns an empty collection
```
DB::table('testing_a.users')->join('testing_b.user_types as ud', 'ud.user_id', '=', 'users.id')->get()
```

My database config file
```
'mysql' => [
            'driver' => 'mysql',
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => env('DB_DATABASE', 'testing_a'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'unix_socket' => env('DB_SOCKET', ''),
            'charset' => 'utf8mb4',
            'collation' => 'utf8mb4_unicode_ci',
            'prefix' => '',
            'strict' => true,
            'engine' => null,
        ],

        'mysql2' => [
            'driver' => 'mysql',
            'host' => env('DB_HOST', '127.0.0.1'),
            'port' => env('DB_PORT', '3306'),
            'database' => env('DB_DATABASE_2', 'testing_b'),
            'username' => env('DB_USERNAME', 'forge'),
            'password' => env('DB_PASSWORD', ''),
            'unix_socket' => env('DB_SOCKET', ''),
            'charset' => 'utf8mb4',
            'collation' => 'utf8mb4_unicode_ci',
            'prefix' => '',
            'strict' => true,
            'engine' => null,
        ],

```

My .env file
```
DB_CONNECTION=mysql
DB_HOST=127.0.0.1
DB_PORT=3306
DB_DATABASE=testing_a
DB_USERNAME=homestead
DB_PASSWORD=secret

DB_DATABASE_2=testing_b
```

My phpunit.xml
```
<env name="DB_CONNECTION" value="mysql"/>
<env name="DB_DATABASE" value="testing_a"/>
<env name="DB_DATABASE_2" value="testing_b"/>
```

### Steps To Reproduce:
I have created a sample project here https://github.com/coxy121/multiple-databases that shows an example of this error in a simple project. 

Once set up, running phpunit will dump out the query results and show one collection and 5 empty collections. 

After running php artisan migrate --seed
Visiting the default route in a browser dumps the results of several queries all returning collections as expected.

**Comment by SCIF** [comment]
Hi @coxy121 , it looks like this issue is [duplicate](https://github.com/laravel/framework/issues/23253#issuecomment-367673503) and not actually the issue.

**Comment by coxy121** [comment]
Hi @SCIF , thanks for the reply.

In my main app I have about 10 different databases, and have multiple relationships between the models on the different databases and these are working without any problems. I also run multiple manual queries joining tables across the various databases with out any issues.

In my model classes I am setting the connection and table name properties.

All the database are on the same connection and use the same host, username and password, and just a different database name.
Therfore, although they are different databases they are all on the same connection, so not cross connection relationships as such, which it looks like that issue is.

Running a query using DB::connection(‘connectionName’)->.... and viewing in a browser I can set the connection to either mysql or mysql2 etc and successfully return the expected results, but when running phpunit would return no results which I have been unable to debug.

**Comment by coxy121** [comment]
I am not sure if this is the expected behavior, but I was able to resolve this by ensuring that during phpunit testing the connection used was always the default connection as specified in the phpunit.xml file. 

As my databases were on the same connection this allowed me to test successfully. 
I did this by creating a trait for each of the additional connections and replacing the following code in each model with the trait.

Initially I had
```
protected $connection = 'mysql2';
```

I replaced this, with a connection specific trait like below
```
trait MySQL2ConnectionTrait
{
    public function __construct(array $attributes = [])
    {
        parent::__construct($attributes);

        $this->connection = app()->environment(['testing']) ? 'mysql' : 'mysql2';

        $this->table = 'testing_b'.$this->getTable();
    }
}
```
